### PR TITLE
Added EntryPoint/MainLoop support (close #3075)

### LINKED ISSUE #3075: Std Main Loop
Each platform comes with its own “main event loop” implementation, making it hard to build crossplatform API over it.

We shoul provide a base abstract loop definition that can help this.

Discussion opened on the topic.

### PR DESCRIPTION
**Pull request by ncannasse**

### CODE REVIEW

**Comment by Simn** [comment]
A workflow remark: It's better to have these branches on ncannasse/haxe because otherwise it clogs up our CI. Every commit causes two CI builds: One for the branch itself and one for the merged version of it. By having the branch on another account these builds are split up nicely.

File: src/typing/typer.ml [hunk 1]
```diff
@@ -4277,7 +4277,20 @@ let get_main ctx =
...
+			let ec = (match et with TClassDecl c -> c | _ -> assert false) in
+			let ef = PMap.find "run" ec.cl_statics in
+			let p = null_pos in
+			let et = mk (TTypeExpr et) (TAnon { a_fields = PMap.empty; a_status = ref (Statics ec) }) p in
+			let call = mk (TCall (mk (TField (et,FStatic (ec,ef))) ef.cf_type p,[])) ctx.t.tvoid p in
+			mk (TBlock [main;call]) ctx.t.tvoid p
+		with Not_found ->
+			main
+		) in
+		Some main
```
Comment (Simn) [inline src/typing/typer.ml]: Sorry for shaking things up again but I really don't like this. Making this depend on DCE is way too implicit and leads to confusing behavior when switching it on and off. I think this whole thing should be opt-in for the user/framework, no implicitly added by the compiler depending the other compilation parameters.

Comment (nadako) [inline src/typing/typer.ml]: +1

Comment (Simn) [inline src/typing/typer.ml]: I misread this initially and didn't realize it was part of Typer.generate which obviously runs _before_ DCE. That makes it slightly better, but I still don't think it's a good solution.

**Comment by ncannasse** [comment]
Any idea it tries to access neko.vm.Lock while I have an explicit `#if (neko && !macro)` ?

**Comment by ncannasse** [comment]
Could someone confirm if the PHP error is related or not ?
(The JS error does not seem to be for instance)

**Comment by Simn** [comment]
Why are we knowingly merging something that fails the tests?

**Pull request merged**